### DISCUSSION TITLE
Reindexing large table - how to avoid timeout errors?

### DISCUSSION THREAD

**Discussion by helloivanco** [supabase/supabase#17821] [Questions]
Hi all, 

I have an issue with being able to index / reindex a table with many records (100k+ rows) 

Creating an index in SQL editor results in a timeout. 
For example running: ```CREATE INDEX on documents USING hnsw (embedding vector_cosine_ops);```

With my limited SQL knowledge, I've also tried to create a simple node script to run the indexing in batches but it's not working properly.

```
import dotenv from 'dotenv';
import pg from 'pg';
const { Client } = pg;

dotenv.config();

const client = new Client({
  connectionString: `postgresql://postgres:${process.env.POSTGRES_PASSWORD}@db.xxxxxxx.supabase.co:5432/postgres`,
});

async function createIndexInBatches(tableName, columnName, batchSize) {
  await client.connect();

  try {
    let lastId = 0;
    let rowCount = 0;

    while (true) {
      const result = await client.query(
        `SELECT id FROM ${tableName} WHERE id > $1 ORDER BY id ASC LIMIT ${batchSize}`,
        [lastId]
      );

      if (result.rowCount === 0) {
        break;
      }

      const ids = result.rows.map((row) => row.id);

      await client.query(`
      CREATE INDEX IF NOT EXISTS documents_embedding_idx
      ON public.documents
      USING hnsw (embedding vector_cosine_ops)
      WHERE id >= ${ids[0]} AND id <= ${ids[ids.length - 1]};
    `);

      rowCount += result.rowCount;
      lastId = ids[ids.length - 1];

      console.log(`Updating index for ${rowCount} rows in ${tableName}`);
    }
  } catch (error) {
    console.error(`Error updating index for ${tableName}`, error);
  } finally {
    client.end();
  }
}

createIndexInBatches('documents', 'embedding', 1000);
```
Pretty sure the issue lies here:
```  
await client.query(`
      CREATE INDEX IF NOT EXISTS documents_embedding_idx
      ON public.documents
      USING hnsw (embedding vector_cosine_ops)
      WHERE id >= ${ids[0]} AND id <= ${ids[ids.length - 1]};
    `);
```

Anyway, I'd appreciate any help troubleshooting this and also how to confirm that the index is properly generated.

There seems to be limited information on indexing / reindexing large Supabase databases which is quite the norm for vector embeddings. Hope this can help others out there as well. 

Thank you so much in advance!

**Comment by danielgwilson** [comment]
Did you ever get this working? I have 1.5M+ rows 💀

**Reply by KKosana318** [reply]
This isn't that large! I was able to index a table with 1 million rows pretty fast with the following strategies:

1. Use a compute add-on. 4XL worked for me
2. Increase the disk limit of your database. 70GB worked for me
3. Upgrade your postgres + pgvector version to enable parallel builds: https://supabase.com/blog/pgvector-fast-builds

**Comment by kevoconnell** [comment]
Hey @charislam !  Maybe you are the wrong person to ask, but is there a good way to do this? currently getting 504 errors when i try to make an index - also getting an upstream error when I add a index with GIN (im doing hybrid search) (I have ~400k rows)

Would really appreciate some pointers on this, want to make sure im not missing something!

**Reply by TheOtherBrian1** [reply]
Hi @MrKevinOConnell, I wrote a [guide](https://github.com/orgs/supabase/discussions/21379) to help with indexing/reindexing. It's specifically for HSNW, but it can be generalized to most cases. I also wrote a [sister-guide](https://github.com/orgs/supabase/discussions/22449) for picking the best indexes.

**Comment by Dizotoff** [comment]
Same problem here, I have 60k+ embeddings. I specified the index name, so I can track if was created or not more easily with.

```
create index if not exists candidates_embedding_index 
on candidates
using hnsw (embedding vector_cosine_ops);
```

I tried to apply it both via supabase sql console and prisma migration, no index was created. Both got timed out.

Update: Tried to apply it via pgAdmin, got following error: 

```
NOTICE:  hnsw graph no longer fits into maintenance_work_mem after 19732 tuples
ERROR:  canceling statement due to statement timeout 

SQL state: 57014
```

Update 2: 

You can see all existing indices for the particular table using. Can be helpful if you're trying to figure out if those exist or don't.

```
select * from pg_indexes where tablename = 'my_table'
```

**Comment by Dizotoff** [comment]
Here's the guide that should help with this issue: https://github.com/orgs/supabase/discussions/21379

**Comment by KKosana318** [comment]
Here's what worked for me, with a table of ~1M rows

1. Use a compute add-on. 4XL worked for me
2. Increase the disk limit of your database. 70GB worked for me
3. Upgrade your postgres + pgvector version to enable parallel builds: https://supabase.com/blog/pgvector-fast-builds
4. Make sure to increase the session timeout to allow enough time for the index to build. Supabase timeout policies are a bit nuanced depending on where you're querying from, so make sure your timeout is _actually_ increased: https://supabase.com/docs/guides/database/postgres/configuration#changing-the-default-timeout
5. Monitor the progress of the build as it runs with the following command (taken from https://github.com/pgvector/pgvector): 
    `SELECT phase, round(100.0 * blocks_done / nullif(blocks_total, 0), 1) AS "%" FROM pg_stat_progress_create_index;`
    Check the progress of your build over time to see if you're coming too close to your session timeout, or if your issue lies somewhere 
    else.
